File: 2720-minimize-the-maximum-difference-of-pairs/minimize-the-maximum-difference-of-pairs.java
class Solution {

    public boolean solve(int mid, int[] nums, int p) {
        int n = nums.length, cnt = 0, i = 0;
      
        while(i < n - 1) {
            if(nums[i+1] - nums[i] <= mid) {
                i += 2;
                cnt++;
            } else i++;

            if(cnt >= p) {
                return true;
            }
        }
        return false;
    }

    public int minimizeMax(int[] nums, int p) {
        Arrays.sort(nums);
        int ans = 0, n = nums.length, l = 0, h = nums[n -1], mid;
        
        
        while(l <= h) {
            mid = l + (h - l)/2;
            if(solve(mid, nums, p))  {
                ans = mid;
                h = mid - 1;
            } else l = mid + 1;
        }

        return ans;
    }
}

File: 2720-minimize-the-maximum-difference-of-pairs/minimize-the-maximum-difference-of-pairs.cpp
class Solution {
public:

    bool solve(int &mid, vector<int>& nums, int &p) {
        int cnt = 0, i = 0, n = nums.size();

        while(i < n - 1) {
            if(nums[i + 1] - nums[i] <= mid) {
                cnt++;
                i += 2;
            } else i++;
            if(cnt >= p) return true;
        }
        return false;
        
    }
    int minimizeMax(vector<int>& nums, int p) {
        sort(nums.begin(), nums.end());
        int n = nums.size(), ans = 0, l = 0, h = nums[n-1], mid;
        // for(int i = 0 ; i < n; i++) cout<<nums[i]<<" ";
        while(l <= h) {
            mid = l + (h - l)/2;
            // cout<<endl<<"mid: " <<mid;

            if(solve(mid, nums, p)) {
                ans = mid;
                h = mid - 1;
            } else {
                l = mid +1;
            }
        }
        
      return ans;
    }
};